library/alloc/src/collections/vec_deque/splice.rs

```diff
@@ -0,0 +1,143 @@
+use core::alloc::Allocator;
+
+use crate::alloc::Global;
+use crate::collections::vec_deque::Drain;
+use crate::vec::Vec;
+
+/// A splicing iterator for `VecDeque`.
+///
+/// This struct is created by [`VecDeque::splice()`][super::VecDeque::splice].
+/// See its documentation for more.
+///
+/// # Example
+///
+/// ```
+/// # #![feature(deque_extend_front)]
+/// # use std::collections::VecDeque;
+///
+/// let mut v = VecDeque::from(vec![0, 1, 2]);
+/// let new = [7, 8];
+/// let iter: std::collections::vec_deque::Splice<'_, _> = v.splice(1.., new);
+/// ```
+#[unstable(feature = "deque_extend_front", issue = "146975")]
+#[derive(Debug)]
+pub struct Splice<
+    'a,
+    I: Iterator + 'a,
+    #[unstable(feature = "allocator_api", issue = "32838")] A: Allocator + 'a = Global,
+> {
+    pub(super) drain: Drain<'a, I::Item, A>,
+    pub(super) replace_with: I,
+}
+
+#[unstable(feature = "deque_extend_front", issue = "146975")]
+impl<I: Iterator, A: Allocator> Iterator for Splice<'_, I, A> {
+    type Item = I::Item;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.drain.next()
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.drain.size_hint()
+    }
+}
+
+#[unstable(feature = "deque_extend_front", issue = "146975")]
+impl<I: Iterator, A: Allocator> DoubleEndedIterator for Splice<'_, I, A> {
+    fn next_back(&mut self) -> Option<Self::Item> {
+        self.drain.next_back()
+    }
+}
+
+#[unstable(feature = "deque_extend_front", issue = "146975")]
+impl<I: Iterator, A: Allocator> ExactSizeIterator for Splice<'_, I, A> {}
+
+#[unstable(feature = "deque_extend_front", issue = "146975")]
+impl<I: Iterator, A: Allocator> Drop for Splice<'_, I, A> {
+    fn drop(&mut self) {
+        // This will set drain.remaining to 0, so its drop won't try to read deallocated memory on
+        // drop.
+        self.drain.by_ref().for_each(drop);
+
+        // At this point draining is done and the only remaining tasks are splicing
+        // and moving things into the final place.
+
+        unsafe {
+            let source_deque = self.drain.deque.as_mut();
+
+            let tail_len = self.drain.tail_len; // #elements behind the drain
+
+            if tail_len == 0 {
+                source_deque.extend(self.replace_with.by_ref());
+                return;
+            }
+
+            if !self.drain.fill(&mut self.replace_with) {
```

Comment:
Also, can you restore the comment from Vec ("First fill the range left by drain().")?